### DISCUSSION TITLE
Use Jquery from twig file

### DISCUSSION THREAD

**Discussion by comsup** [flyntwp/flynt#478] [Q&A] [answered]
Hello,
I need to update a function writed in the script.js file (in my module) from my twig (generation of informations from the database) but when I'm trying to use jquery I got the error Uncaught ReferenceError: $ is not defined. This error is when I tri to use jquery from the twig, jquery is well load in the script.js from import. Is there a way to make jquery working in the twig file? Or write the data in the script.js from the functions.js? Thanks!

**Comment by comsup** [comment]
I found a solution but may be you have something better, I write a Js array in my twig with all datas I need from the backend and after I make a reference to it on the script.js where jquery is working.

**Comment by aaronmeder** [comment]
Hey @comsup 

Hmm not totally sure I understand what you are trying to achieve...can you elaborate a bit more on what it is that you try to do and show the code of your current solution?

Aaron

**Comment by comsup** [comment]
I'm making a map with Mapael (https://www.vincentbroute.fr/mapael/) with custom svg. The map is loaded from my script.js from the module, where I import jquery but I want to add plot on this map, and the plot are made from the Wp backend so they are generated on the twig file. So in my twig I get the informations of the plots : 
  <script>
    **var newPlots** = {
    {% for reg in region  %}
      {% for plot in reg.bloc_depot %}
      '{{ plot.blocID }}': {
        latitude: {{ plot.latitude }},
        longitude: {{ plot.longitude }},
      },
      {% endfor %}
    {% endfor %}
    }
  </script>
and in the script I load the Map with this var (newPlots) :



function init () {
    const $document = $(document)
    const mapaEl = $('.mapcontainer')
    $document.ready(function(){
        mapaEl.mapael({
            map: {
                name: "france_custom"
                ,
                defaultArea: {
                    attrs: {
                        fill: "#ededec",
                        stroke: "#638998",
                        cursor: "pointer"
                    },
                    attrsHover: {
                        animDuration: 350,
                        fill: "#d9d9d9",
                    },
                    eventHandlers: {
                        click: function (e, id, mapElem, textElem) {
                            var newData = {
                                'areas': {}
                            };
                            for(var i =1; i<15; i++){
                                newData.areas["region-"+i] = {
                                    attrs: {
                                        fill: "#ededec"
                                    }
                                };
                            }
                            if (mapElem.originalAttrs.fill == "#ededec") {
                                newData.areas[id] = {
                                    attrs: {
                                        fill: "#1360aa"
                                    }
                                };
                            } else {
                                newData.areas[id] = {
                                    attrs: {
                                        fill: "#ededec"
                                    }
                                };
                            }
                            $("#"+id).css({display: 'block'}).fadeIn('slow');
                            $(".mapcontainer").trigger('update', [{mapOptions: newData}]);
                        }
                    },
                },
            },
            **// ici on fait référence
            plots: newPlots**
        });
    });
}

$(init)





 but the code is plit between these 2 files and I was wondering if there is a better way to achieve this.

**Comment by aaronmeder** [comment, marked as answer]
What you could do is getting & preparing the data in functions.php, then passing it on the the twig template where you use json_encode to output that data into a script tag. After that use that information from jQuery in script.js.

Here is a minimal demo:

**functions.php**
```php
<?php

namespace Flynt\Components\BlockTest;

use Flynt\FieldVariables;

add_filter('Flynt/addComponentData?name=BlockTest', function ($data, $componentName) {
    // get data, depends on how you stored it...
    // just setting it here for the example
    $plots = ['plot1', 'plot2', 'plot3'];

    // prepare data to be passed to twig file
    $data['jsonData'] = [
        'plots' => $plots,
    ];

    return $data;
}, 10, 2);



function getACFLayout()
{
    return [
        'name' => 'BlockTest',
        'label' => 'Test',
        'sub_fields' => [
            [
                'label' => '',
                'name' => 'testMessage',
                'type' => 'message',
                'message' => 'Only a test component',
                'new_lines' => 'wpautop',
                'esc_html' => 1
            ],
        ],
    ];
}

```

**index.twig**
```twig
<section is="flynt-block-test" class="flyntComponent">
  <script type="application/json">{{ jsonData|json_encode }}</script>
  <div class="map">
   {# Will be initialized later from JS #}
  </div>
</section>
```


**script.js**
```js
import $ from 'jquery'

class BlockTest extends window.HTMLElement {
  constructor (...args) {
    const self = super(...args)
    self.init()
    return self
  }

  init () {
    this.$ = $(this)
    this.props = this.getInitialProps()
    this.resolveElements()
  }

  /* This gets the data from the script tag in twig and makes it accessible for you to use here */
  getInitialProps () {
    let data = {}
    try {
      data = JSON.parse($('script[type="application/json"]', this).text())
    } catch (e) {}
    return data
  }

  resolveElements () {
    // get elements
  }

  connectedCallback () {
    // initialize your map
    console.log('Your Plots: ', this.props.plots)
  }
}

window.customElements.define('flynt-block-test', BlockTest, {
  extends: 'section'
})

```

You can see this pattern in places like the [SliderImages](https://github.com/flyntwp/flynt/tree/master/Components/SliderImages).

Is this what you were looking for?

Greetings from Switzerland 🇨🇭
Aaron

**Comment by comsup** [comment]
Yes many thanks for this very explicit reply!

**Reply by aaronmeder** [reply]
Happy it helped 🙌🏻
Please mark my reply as the answer to your question, so that other people might find it easier.

Enjoy developing with Flynt!
